packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,420 @@
+    event ProposalDeleted(uint256 indexed rewardIndex, uint256 indexed proposalIndex);
+    event MerkleDistributorSet(address indexed merkleDistributor);
+
+    /**
+     * @notice Constructor.
+     * @param _bondToken ERC20 token that the bond is paid in.
+     * @param _finder Finder to look up UMA contract addresses.
+     */
+    constructor(FinderInterface _finder, IERC20 _bondToken) {
+        finder = _finder;
+        require(_getCollateralWhitelist().isOnWhitelist(address(_bondToken)), "Bond token not supported");
+        bondToken = _bondToken;
+        syncUmaEcosystemParams();
+    }
+
+    /********************************************
+     *            FUNDING FUNCTIONS             *
+     ********************************************/
+
+    /**
+     * @notice Allows any caller to create a rewards object and deposit tokens that are linked to these rewards.
+     * @dev The caller must approve this contract to transfer `maximumRewardAmount` amount of `rewardToken`.
+     * @param rewardToken ERC20 token that the rewards will be paid in.
+     * @param maximumRewardAmount Maximum reward amount that the sponsor is posting for distribution.
+     * @param earliestProposalTimestamp Starting timestamp when proposals for distribution can be made.
+     * @param priceIdentifier Identifier that should be passed to the Optimistic Oracle on proposed distribution.
+     * @param customAncillaryData Custom ancillary data that should be sent to the Optimistic Oracle on proposed
+     * distribution.
+     * @param optimisticOracleProposerBond Amount of bondToken that should be posted in addition to final fee
+     * to the Optimistic Oracle on proposed distribution.
+     * @param optimisticOracleLivenessTime Liveness period in seconds during which proposed distribution can be
+     * disputed through Optimistic Oracle.
+     */
+    function createReward(
+        uint256 maximumRewardAmount,
+        uint256 earliestProposalTimestamp,
+        uint256 optimisticOracleProposerBond,
+        uint256 optimisticOracleLivenessTime,
+        bytes32 priceIdentifier,
+        IERC20 rewardToken,
+        bytes calldata customAncillaryData
+    ) external nonReentrant() {
+        require(address(merkleDistributor) != address(0), "Missing MerkleDistributor");
+        require(_getIdentifierWhitelist().isIdentifierSupported(priceIdentifier), "Identifier not registered");
+        require(_ancillaryDataWithinLimits(customAncillaryData), "Ancillary data too long");
+        require(optimisticOracleLivenessTime >= MINIMUM_LIVENESS, "OO liveness too small");
+        require(optimisticOracleLivenessTime < MAXIMUM_LIVENESS, "OO liveness too large");
+
+        // Pull maximum rewards from the sponsor.
+        rewardToken.safeTransferFrom(msg.sender, address(this), maximumRewardAmount);
+
+        // Store funded reward and log created reward.
+        Reward memory reward =
+            Reward({
+                sponsor: msg.sender,
```

Comment:
does the key "sponsor" make the most sence here? would "proposer" be better than "sponsor"? I bring this up when looking at the `increaseReward` method where you check the value of `sponsor` not being the 0 address and that's a bit confusing as there can then be multiple `sponsors`. WDYT?

Reply:
"proposer" is the party submitting merkleRoot for proposed reward payout distribution together with proposal bond + final fee, and it would be recorded as "proposer" in OO price proposal. So it would be confusing to use that naming in this contract for different role.

Here at OptimisticDistributor "sponsor" is the first party funding the reward and it is unique per Reward struct. I agree it could be bit confusing as there could be multiple sponsors funding through `increaseReward` method that currently are not recorded. Not sure if worth tracking all of them in state variables, but maybe I should emit add-on sponsors in the `RewardIncreased` events. Then the price identifier/ancillary data could have instructions on how to distribute any remaining rewards (e.g. if used in bribing as rewards per % gauge allocation) back to sponsors as part of merkle root distribution. Price identifier/ancillary data could then also decide if all remaining rewards go back to first sponsor calling `createReward` method or distributed pro-rata to all sponsors also considering all callers of `increaseReward` for the same `rewardIndex`.

Now I realize this contract currently does not strictly require storing original sponsor as part of Reward struct and one could check the validity of `rewardIndex` by non-zero `rewardToken` address instead. But there might be future upgrades that allow clawing back unclaimed rewards from MerkleDistributor if predefined time has passed (e.g. due to data availability concerns for storing IPFS payout object) - then OptimisticDistributor might need original sponsor address stored to return those unclaimed rewards.

Reply:
I think having `proposer` mean different things for this contract and the OO is fine, as is the case for the across-v2 contracts

Reply:
I would still believe "sponsor" better represents a party funding the rewards and "proposer" being a party proposing distribution of those rewards

Reply:
@Reinis-FRP I think the point that matters, like you called out, is they are the one _sponsoring_ the payout so it works just fine.